action/protocol/execution/protocol_test.go

```diff
@@ -41,6 +42,149 @@ const (
 	testTriePath = "trie.test"
 )
 
+type execCfg struct {
+	executor            string
+	privateKey          keypair.PrivateKey
+	codeHex             string
+	amount              uint64
+	gasLimit            uint64
+	gasPrice            int64
+	hasReturnValue      bool
+	expectedReturnValue interface{}
+	expectedBalances    map[string]*big.Int
+}
+
+type smartContractTest struct {
+	prepare map[string]*big.Int
+	deploy  execCfg
+	// the order matters
+	executions []execCfg
+}
+
+func runExecution(
+	bc blockchain.Blockchain,
+	ecfg *execCfg,
+	contractAddr string,
+) (*action.Receipt, error) {
+	code, err := hex.DecodeString(ecfg.codeHex)
+	if err != nil {
+		return nil, err
+	}
+	nonce, err := bc.Nonce(ecfg.executor)
+	if err != nil {
+		return nil, err
+	}
+	exec, err := action.NewExecution(
+		ecfg.executor,
+		contractAddr,
+		nonce+1,
+		big.NewInt(int64(ecfg.amount)),
+		ecfg.gasLimit,
+		big.NewInt(ecfg.gasPrice),
+		code,
+	)
+	if err != nil {
+		return nil, err
+	}
+	builder := &action.EnvelopeBuilder{}
+	elp := builder.SetAction(exec).
+		SetDestinationAddress(exec.DstAddr()).
+		SetNonce(exec.Nonce()).
+		SetGasLimit(exec.GasLimit()).
+		Build()
+	selp, err := action.Sign(elp, exec.Executor(), ecfg.privateKey)
+	if err != nil {
+		return nil, err
+	}
+	blk, err := bc.MintNewBlock(
+		[]action.SealedEnvelope{selp},
+		ta.Addrinfo["producer"],
+		nil,
+		nil,
+		"",
+	)
+	if err != nil {
+		return nil, err
+	}
+	if err := bc.ValidateBlock(blk, true); err != nil {
+		return nil, err
+	}
+	if err := bc.CommitBlock(blk); err != nil {
+		return nil, err
+	}
+
+	return bc.GetReceiptByExecutionHash(exec.Hash())
+}
+
+func (sct *smartContractTest) run(r *require.Assertions) {
+	// prepare blockchain
```

Comment:
nit: run() can be splited into smaller funcs like preparebc(), deploy(), execToExpect().